src/Lykke.Job.CandlesHistoryWriter.Services/HistoryMigration/CandlesFiltrationManager.cs

```diff
@@ -0,0 +1,146 @@
+            {
+                priceTypeTasks.Add(
+                    Task.Run(() => 
+                        DoFiltrateAsync(request.AssetPairId, request.LimitLow, request.LimitHigh, priceType, epsilon, analyzeOnly)
+                            .GetAwaiter()
+                            .GetResult()));
+            }
+
+            Task.WhenAll(priceTypeTasks.ToArray()).ContinueWith(t =>
+            {
+                Health.State = CandlesFiltrationState.Finished;
+
+                if (analyzeOnly)
+                    _log.WriteInfo(nameof(CandlesFiltrationManager), nameof(Filtrate),
+                        $"Filtration for {request.AssetPairId} finished: analyze only. Total amount of candles to delete: {Health.DeletedCandlesCount.Values.Sum()}, " +
+                        $"total amount of candles to replace: {Health.ReplacedCandlesCount.Values.Sum()}. Errors count: {Health.Errors.Count}.");
+                else
+                    _log.WriteInfo(nameof(CandlesFiltrationManager), nameof(Filtrate),
+                        $"Filtration for {request.AssetPairId} finished. Total amount of deleted Sec candles: {Health.DeletedCandlesCount.Values.Sum()}, " +
+                        $"total amount of replaced bigger candles: {Health.ReplacedCandlesCount.Values.Sum()}. Errors count: {Health.Errors.Count}.");
+            });
+
+            return FiltrationLaunchResult.Started;
+        }
+
+        private async Task DoFiltrateAsync(string assetId, double limitLow, double limitHigh, CandlePriceType priceType, double epsilon, bool analyzeOnly)
+        {
+            try
+            {
+                await _log.WriteInfoAsync(nameof(CandlesFiltrationManager), nameof(DoFiltrateAsync),
+                    $"Starting candles with extreme prices filtration for price type {priceType}...");
+
+                // First of all we need to find out if there are any extreme candles with the given parameters or there are not.
+                var extremeCandles = await _candlesFiltrationService.TryGetExtremeCandlesAsync(assetId, priceType, limitLow, limitHigh, epsilon);
+
+                if (extremeCandles == null)
+                {
+                    await _log.WriteInfoAsync(nameof(CandlesFiltrationManager), nameof(DoFiltrateAsync),
+                        $"There are no extreme price candles for price type {priceType}. Skipping.");
+                    return;
+                }
+
+                // If we were asked just to analyze the amount of troubled candles, we need to prepare the answer here
+                // and exit.
+                if (analyzeOnly)
+                {
+                    var secondCandlesCount = extremeCandles
+                        .Count(c => c.TimeInterval == CandleTimeInterval.Sec);
+
+                    Health.DeletedCandlesCount[priceType] = secondCandlesCount;
+                    Health.ReplacedCandlesCount[priceType] = extremeCandles.Count - secondCandlesCount;
```

Comment:
Yes, it can be. And it actually is. Such a counting principle provides only an estimated amounts. But we can not calculate the accurate ones until we actually perform a filtration.
The reason why I still consider this approach good enougth is that the summ of "deleted" and "replaced" candles will (should) be generally "true".